### PR TITLE
[Merged by Bors] - feat: port Analysis.Normed.Group.Seminorm

### PR DESCRIPTION
**Pull request by j-loreaux**
- [x] depends on: #2401
- [x] depends on: #2419

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2401
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 1]
```diff
@@ -0,0 +1,1020 @@
...
+end AddCommGroup
+
+end NonarchAddGroupSeminorm
+
+namespace GroupSeminorm
+
+variable [Group E] [SMul R ℝ] [SMul R ℝ≥0] [IsScalarTower R ℝ≥0 ℝ]
+
+-- porting note: added manually because of problems with `to_additive`
+instance _root_.AddGroupSeminorm.toOne [DecidableEq E] [AddGroup E] : One (AddGroupSeminorm E) :=
```
Comment (fpvandoorn) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:630]: This is the same instance as on line 452, right?

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+      mul_le' := fun _ _ =>
+        by
+        simp only [← smul_one_smul ℝ≥0 r (_ : ℝ), NNReal.smul_def, smul_eq_mul]
+        exact
+          (mul_le_mul_of_nonneg_left (map_mul_le_add p _ _) <| NNReal.coe_nonneg _).trans_eq
+            (mul_add _ _ _)
+      inv' := fun x => by simp_rw [map_inv_eq_map p] }⟩
+
+-- porting note: `to_additive` tried to change `SMul` to `VAdd` so we add this manually
+instance _root_.AddGroupSeminorm.isScalarTower [AddGroup E] [SMul R' ℝ] [SMul R' ℝ≥0]
```
Comment (fpvandoorn) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:677]: This instance is also duplicated

File: Mathlib/Analysis/Normed/Group/Seminorm.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/Analysis/Normed/Group/Seminorm.lean]: whoops, thank for spotting that. They were so far away from each other that I didn't notice.

**Comment by urkud** [comment]
What is the current status of this PR? I have most of the other dependencies of `analysis.normed.group.basic` done (awaiting review).

**Comment by j-loreaux** [comment]
Just waiting on the `to_additive` dependent PR to be reviewed and approved.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2401~~
* ~~leanprover-community/mathlib4#2419~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by urkud** [comment]
Could you please merge or rebase?

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 3]
```diff
@@ -0,0 +1,1001 @@
...
+@[to_additive (attr := simp)]
+theorem add_apply (x : E) : (p + q) x = p x + q x :=
+  rfl
+#align group_seminorm.add_apply GroupSeminorm.add_apply
+#align add_group_seminorm.add_apply AddGroupSeminorm.add_apply
+
+-- TODO: define `SupSet` too, from the skeleton at
+-- https://github.com/leanprover-community/mathlib/pull/11329#issuecomment-1008915345
+@[to_additive]
+instance : HasSup (GroupSeminorm E) :=
```
Comment (urkud) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:296]:
```suggestion
instance : Sup (GroupSeminorm E) :=
```

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 4]
```diff
@@ -0,0 +1,1001 @@
...
+    BddBelow (range fun y => p y + q (x / y)) :=
+  ⟨0, by
+    rintro _ ⟨x, rfl⟩
+    dsimp
+    exact add_nonneg (map_nonneg _ _) (map_nonneg _ _)⟩ -- porting note: was `positivity`
+#align group_seminorm.mul_bdd_below_range_add GroupSeminorm.mul_bddBelow_range_add
+#align add_group_seminorm.add_bdd_below_range_add AddGroupSeminorm.add_bddBelow_range_add
+
+@[to_additive]
+noncomputable instance : HasInf (GroupSeminorm E) :=
```
Comment (urkud) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:406]:
```suggestion
noncomputable instance : Inf (GroupSeminorm E) :=
```

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 5]
```diff
@@ -0,0 +1,1001 @@
...
+theorem zero_apply (x : E) : (0 : NonarchAddGroupSeminorm E) x = 0 :=
+  rfl
+#align nonarch_add_group_seminorm.zero_apply NonarchAddGroupSeminorm.zero_apply
+
+instance : Inhabited (NonarchAddGroupSeminorm E) :=
+  ⟨0⟩
+
+-- TODO: define `SupSet` too, from the skeleton at
+-- https://github.com/leanprover-community/mathlib/pull/11329#issuecomment-1008915345
+instance : HasSup (NonarchAddGroupSeminorm E) :=
```
Comment (urkud) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:584]:
```suggestion
instance : Sup (NonarchAddGroupSeminorm E) :=
```

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 6]
```diff
@@ -0,0 +1,1001 @@
...
+
+@[to_additive (attr := simp)]
+theorem add_apply (x : E) : (p + q) x = p x + q x :=
+  rfl
+#align group_norm.add_apply GroupNorm.add_apply
+#align add_group_norm.add_apply AddGroupNorm.add_apply
+
+-- TODO: define `SupSet`
+@[to_additive]
+instance : HasSup (GroupNorm E) :=
```
Comment (urkud) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:846]:
```suggestion
instance : Sup (GroupNorm E) :=
```

File: Mathlib/Analysis/Normed/Group/Seminorm.lean [hunk 7]
```diff
@@ -0,0 +1,1001 @@
...
+#align nonarch_add_group_norm.coe_le_coe NonarchAddGroupNorm.coe_le_coe
+
+@[simp, norm_cast]
+theorem coe_lt_coe : (p : E → ℝ) < q ↔ p < q :=
+  Iff.rfl
+#align nonarch_add_group_norm.coe_lt_coe NonarchAddGroupNorm.coe_lt_coe
+
+variable (p q) (f : F →+ E)
+
+instance : HasSup (NonarchAddGroupNorm E) :=
```
Comment (urkud) [inline Mathlib/Analysis/Normed/Group/Seminorm.lean:967]:
```suggestion
instance : Sup (NonarchAddGroupNorm E) :=
```

**Review by dupuisf — COMMENTED** [review]
Thanks!

bors r+

**Pull request closed without merging**